src/consumer_queue.cpp

```diff
@@ -23,10 +22,13 @@ consumer_queue::~consumer_queue() {
 }
 
 void consumer_queue::push_sample(const sample_p &sample) {
+	// acquire lock for more predictable behavior in regards to pop_sample()
+	std::unique_lock<std::mutex> lk(lock_);
 	while (!buffer_.push(sample)) {
 		sample_p dummy;
 		buffer_.pop(dummy);
 	}
```

Comment:
Wrapping this whole section (including the lock) in a code block would resolve the thread safety issue and double free error. Will only need a std::lock_guard then. After the other change to pop_sample, we should be able to leave the notify_one outside the lock as per your suggestion.